frontend/packages/dev-console/src/components/helm/list/HelmReleaseList.tsx

```diff
@@ -1,43 +1,162 @@
   namespace: string;
   secrets?: FirehoseResult;
 }
 
 const HelmReleaseList: React.FC<HelmReleaseListProps> = ({ namespace, secrets }) => {
-  const memoizedSecrets = useDeepCompareMemoize(secrets?.data);
+  const [releasesLoaded, setReleasesLoaded] = React.useState<boolean>(false);
+  const [loadError, setLoadError] = React.useState<string>();
+  const [releases, setReleases] = React.useState([]);
+  const [filteredReleases, setFilteredReleases] = React.useState([]);
+  const secretsCountRef = React.useRef<number>(0);
+  const newCount = secrets.data?.length ?? 0;
+
+  React.useEffect(() => {
+    let destroyed = false;
+    if (secrets.loaded && !secrets.loadError) {
+      if (newCount === 0) {
+        setReleases([]);
+        setLoadError(null);
+        setReleasesLoaded(true);
+      } else if (secretsCountRef.current !== newCount) {
+        setReleasesLoaded(false);
+        fetchHelmReleases(namespace)
+          .then((helmReleases) => {
+            const queryArgument = getQueryArgument('rowFilter-helm-release-status');
+            const activeFilters = queryArgument?.split(',');
+            if (!destroyed) {
+              setReleases(helmReleases);
+              setReleasesLoaded(true);
+              setLoadError(null);
+              if (activeFilters) {
+                setFilteredReleases(filterHelmReleasesByStatus(helmReleases, activeFilters));
+              } else {
+                setFilteredReleases(helmReleases);
+              }
+            }
+          })
+          .catch((err) => {
+            if (!destroyed) {
+              setReleasesLoaded(true);
+              setLoadError(err.message || 'Unable to load Helm Releases');
+            }
+          });
+      }
+      secretsCountRef.current = newCount;
+    }
+    return () => {
+      destroyed = true;
+    };
+  }, [namespace, newCount, secrets.loadError, secrets.loaded]);
+
+  const applyRowFilter = React.useCallback(
+    (filter) => {
+      setFilteredReleases(filterHelmReleasesByStatus(releases, filter));
+    },
+    [releases],
+  );
+
+  const applyTextFilter = React.useCallback(
+    (filter) => {
+      setFilteredReleases(filterHelmReleasesByStatus(releases, filter));
+    },
+    [releases],
+  );
+
+  if (!secrets.loaded || !releasesLoaded) {
+    return <LoadingBox />;
+  }
+
+  if (secrets.loadError || loadError) {
+    return <StatusBox loaded loadError={secrets.loadError || loadError} label="Helm Releases" />;
+  }
 
-  const getHelmReleases = (): Promise<HelmRelease[]> => {
-    return fetchHelmReleases(namespace);
-  };
+  if (releases && releases.length === 0) {
+    const helmImage = () => (
+      <img
+        className="odc-helm-release__empty-list__image"
+        src={getImageForIconClass('icon-helm')}
+        alt=""
+      />
+    );
+    const installURL = { pathname: `/catalog/ns/${namespace}`, search: '?kind=%5B"HelmChart"%5D' };
+    return (
+      <EmptyState variant={EmptyStateVariant.full}>
```

Comment:
We can move the logic of showing `EmptyState` to `CustomResourceList` and make it generic so that other list pages using it would get the same functionality. We can pass `emptyState: React.ReactNode` prop to the list component and it shows that based on the fetched resources. Everything else should work fine since it already re-fetches resources based on the `dependantResource` which is `secrets` in this case. WDYT?

Reply:
The `CustomResourceList` does not take a `FirehoseResult` as a `dependantResource` (the other usage does not use one). If it did we could handle the loading state and empty states much better. I tried doing it that way but there are race conditions with the loading state. Only other thing to do would be to pass the loading state onto the `CustomResourceList` and that seemed to be too much delta and defeats the purpose of the common component.

Reply:
@rohitkrai03 OK, I revisited and updated to continue to use `CustomResourceList`. Added parameters for an `EmptyMsg` and `loaded` flag.